### PR TITLE
Track Table Column Changes (LEAN-3715)

### PR DESCRIPTION
**Pull request by asouqi**
As we know there is no node for the table column, so when we delete a column the changes will not be grouped it's going to be a list of cells deleted, and to make it easy for user to deal with that changes will group them in a column-change that will be added to table `dataTracked`. also that will help us to delegate the update of colspan to prosemirror table command so when the user accepts changes will run removeColumn command and it's going to figure out new colspan value.

### CODE REVIEW

**Review by mbartenev-atypon — CHANGES_REQUESTED** [review]

File: src/ChangeSet.ts [hunk 1]
```diff
@@ -62,9 +63,19 @@ export class ChangeSet {
...
     const rootNodes: TrackedChange[] = []
+    const columnChanges = new Map<string, number>()
     let currentNodeChange: NodeChange | undefined
-    this.changes.forEach((c) => {
-      if (
+    this.changes.forEach((c, index) => {
+      if (c.type === 'column-change') {
+        columnChanges.set(c.dataTracked.id, index)
+        c.children = []
+      } else if (c.dataTracked.column_change_id && columnChanges.has(c.dataTracked.column_change_id)) {
```
[the remark this answers is not in the document]
Comment (asouqi) [reply src/ChangeSet.ts]: updated

File: src/types/change.ts [hunk 2]
```diff
@@ -58,6 +59,12 @@ export type NodeChange = Change & {
   nodeType: string
   children: TrackedChange[]
 }
+export type ColumnChange = Change & {
+  type: 'column-change'
+  nodeType: string
+  tablePosition: number
```
[the remark this answers is not in the document]
Comment (asouqi) [reply src/types/change.ts]: removed, and updated `node-change` type to handle the case of being a reference change

File: src/compute/nodeHelpers.ts [hunk 3]
```diff
@@ -101,3 +106,54 @@ export function getMergeableMarkTrackedAttrs(
...
+  return node.attrs.dataTracked?.find((attr: TrackedAttrs) => attr.column_change_id)
+}
+
+export function addTableColumnChange(
+  tr: Transaction,
+  newTr: Transaction,
+  selection: Selection,
+  emptyAttrs: NewEmptyAttrs
+) {
+  const tableColumnChange = tr.getMeta('tableColumnChange')
```
[the remark this answers is not in the document]
Comment (asouqi) [reply src/compute/nodeHelpers.ts:120]: we can look at the steps if they are all about adding/removing table_cell, but do we have a guarantee that we will not have the same case for other operations in the table?

[the remark this answers is not in the document]
Comment (mbartenev-atypon) [reply src/compute/nodeHelpers.ts:120]: I would say it is possible to sieve out that change but since it's the handling of a direct user input, that should be ok.

File: package.json [hunk 4]
```diff
@@ -63,10 +63,13 @@
...
     "prosemirror-state": "^1.4.2",
+    "prosemirror-utils": "^0.9.6",
     "prosemirror-transform": "^1.7.0",
+    "prosemirror-tables": "^1.3.5",
     "prosemirror-view": "^1.29.1"
   },
   "resolutions": {
-    "eslint-mdx": "2.0.5"
+    "eslint-mdx": "2.0.5",
+    "prosemirror-state": "^1.4.2"
```
Comment (mbartenev-atypon) [inline package.json:73]: why is that needed?

File: src/ChangeSet.ts [hunk 5]
```diff
@@ -62,9 +62,19 @@ export class ChangeSet {
    */
   get changeTree() {
     const rootNodes: TrackedChange[] = []
+    const referenceChanges = new Map<string, number>()
```
Comment (mbartenev-atypon) [inline src/ChangeSet.ts:65]: please provide a comment here so that'd be clearer in future what it's for

File: package.json (hunk 4 above)
Comment (asouqi) [reply package.json:73]: `prosemirror-tables` use old version of state

File: src/ChangeSet.ts (hunk 5 above)
Comment (asouqi) [reply src/ChangeSet.ts:66]: updated

**Review by mbartenev-atypon — APPROVED** [review]
